cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/DistributedStreamService.java

```diff
@@ -286,4 +395,104 @@ public void finished(Throwable failureCause) {
       }
     }
   }
+
+  /**
+   * Aggregate the sizes of all stream writers. A notification is published if the aggregated
+   * size is higher than a threshold.
+   */
+  private final class Aggregator extends AbstractScheduledService {
+
+    private final Map<Integer, Long> streamWriterSizes;
+    private final NotificationFeed streamFeed;
+    private final AtomicLong streamBaseCount;
+    private final Cancellable truncationSubscription;
+    private ScheduledExecutorService executor;
+
+    protected Aggregator(String streamName, long baseCount) {
+      this.streamWriterSizes = Maps.newHashMap();
+      this.streamBaseCount = new AtomicLong(baseCount);
+      this.streamFeed = new NotificationFeed.Builder()
+        .setNamespace(Constants.DEFAULT_NAMESPACE)
+        .setCategory(Constants.Notification.Stream.STREAM_FEED_CATEGORY)
+        .setName(streamName)
+        .build();
+
+      this.truncationSubscription = getStreamCoordinatorClient().addListener(streamName, new StreamPropertyListener() {
```

Comment:
I see, I will make those changes. But still I think the ``Aggregator`` should still be a service, because it needs to perform the aggregation of heartbeats at a fixed rate. Or is that simpler to make ``Aggregator`` a ``Runnable`` and use a ``SchedulingExecutorService`` on it?